### ISSUE TITLE
[9.0][10.0] account_tax_balance performance

### ISSUE THREAD

**Issue by sbidoul** [OCA/account-financial-reporting#292]
I notice account_tax_balance is very slow even when there are few account moves.

A little bit of sql tracing show it repeats this pattern for each tax:
* select all move lines for that tax
* perform additional filtering on move lines

```sql
SELECT "account_move_line_id" FROM "account_move_line_account_tax_rel" WHERE "account_tax_id" IN (383);

SELECT "account_move_line".id FROM "account_move" as "account_move_line__move_id","account_move_line" 
WHERE ("account_move_line"."move_id"="account_move_line__move_id"."id") 
AND (((((("account_move_line"."date" <= '2017-12-31')  
AND  ("account_move_line"."date" >= '2017-01-01'))  
AND  ("account_move_line"."company_id" = 1))  AND  ("account_move_line__move_id"."state" in ('posted')))  
AND  ("account_move_line"."id" in (1,2,3,4,5,6,7,8,9,10,11,12,13,15,16,17,18,19,21,22,23,24,25,26,27,29,30,31,42,43,44,45,46,47,48,60,62,63,64,65,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,89,90,91,93,94,95,97,98,107,108,109,110,11
1,112,113,115,116,118,119,120,121,122,123,124,129,130,137,138,139,140,141,143,144,145,146,147,149,150,151,152,153,154,155,157,158,159,160,162,163,165,166,167,168,169,170,172,173,175,176,177,178,179,180,181,182,183,184,185,186,187,189,190,191,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,215,216,217,218,219,220,221,22
2,223,224,225,226,227,228,229,230,231,232,233,234,235)))  
AND  ("account_move_line__move_id"."move_type" in ('receivable_refund','payable_refund'))) 
ORDER BY "account_move_line"."date" DESC,"account_move_line"."id" DESC,
```

I guess it's due to the many2many relation between move lines and taxes which is not optimizable with auto-join.

**Comment by sbidoul** [comment]
@eLBati do you see the same problem? Any idea on how to optimize this?

**Comment by sbidoul** [comment]
And this query pattern is repeated for each tax (we have ~400 in Luxemburg...).

**Comment by sbidoul** [comment]
Another important performance hit is due to `_search_balance_field` which is used to filter out lines with no data. This could be improved by doing a preliminary search on taxes which are actually used in the reporting period (even if the balance is 0).

**Comment by sbidoul** [comment]
I implemented my comment above in #293. The query pattern in the issue description is still present, but only for taxes actually used, so it's a bit better.

**Comment by sbidoul** [comment]
Experimenting an Odoo improvement that should fix this and many more: https://github.com/odoo/odoo/pull/16066

**Comment by sbidoul** [comment]
Resolved with #293 and odoo/odoo#16066
